FEATURE: Window Scrolling API (Offset/Keyset pagination) [Spring Data JDBC]

chrshnv opened the pull request:
This PR introduces an implementation of Window Scrolling API for Spring Data JDBC, providing efficient, index-friendly pagination over ordered datasets.
Unlike traditional offset pagination, the new API uses keyset (window) navigation, which enables constant-time performance even on large tables.

# Example query:
```sql
WHERE ("DUMMY_ENTITY"."ID_PROP" >= :id_prop AND ("DUMMY_ENTITY"."ID_PROP" > :id_prop1 OR ("DUMMY_ENTITY"."POINT_IN_TIME" > :point_in_time))) ORDER BY "DUMMY_ENTITY"."ID_PROP" ASC, "DUMMY_ENTITY"."POINT_IN_TIME" ASC LIMIT 2
```

Commit:
Scrolling API support #2149

Signed-off-by: Artemiy Degtyarev <chereshnyabtw@icloud.com>
Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

add: StatementFactory new mode for scroll api

Signed-off-by: Artemiy Degtyarev <chereshnyabtw@icloud.com>
Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

add: basic keyset pagination support (without directions)

Signed-off-by: Artemiy Degtyarev <chereshnyabtw@icloud.com>
Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

add: test with two keys

Signed-off-by: Artemiy Degtyarev <chereshnyabtw@icloud.com>
Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

add: sorting for keys not in query

Signed-off-by: Artemiy Degtyarev <chereshnyabtw@icloud.com>
Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

add: limit support

Signed-off-by: Artemiy Degtyarev <chereshnyabtw@icloud.com>
Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

fix: more optimal pg query

Signed-off-by: Artemiy Degtyarev <chereshnyabtw@icloud.com>
Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

fix: remove second compare for one-key query

Signed-off-by: Artemiy Degtyarev <chereshnyabtw@icloud.com>
Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

add: me in headers!

Signed-off-by: Artemiy Degtyarev <chereshnyabtw@icloud.com>
Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

code: move to 'ReflectionUtils'

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

add: unit-test for two key query creation

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

fix: fix query generation for three or more keys

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

documentation

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

fix: query test fix

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

fix: invalid next scroll position building due to difference in property and column name

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

fix: remove unexpected sort creation when column already in sort

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

fix: use RelationalPersistentProperty.getName()
instead of RelationalPersistentProperty.getColumnName().getReference()

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

fix: use RelationalPersistentProperty.getName()
instead of RelationalPersistentProperty.getColumnName().getReference()

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

test: use property name instead of database column

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

fix: getColumnName.getReference -> getName

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

code: more beautiful query building

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

code: fix formatting

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

Fix: offset scrolling - calculate query offset by page size

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

Test: add tests for window keyset position after first page

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

chrshnv commented:
i guess done. awaiting your reviews!

chrshnv commented:
I have a question about the KeysetScrollPosition specification. Should users use property names from POJO or column names from the database?

chrshnv committed:
Polishing.

Reformat keys extraction code.

See #2149

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

Polishing.

Reformat 'applyScrollOrderBy'.

See #2149

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

mp911de commented:
Thanks for your pull request. This is interesting but we don't have currently the bandwidth to evolve JDBC support. We will come back to this one next year or so.

chrshnv committed:
Merge branch 'main' into main

Signed-off-by: Artemiy Degtyarev <chereshnevvv4real@gmail.com>

chrshnv committed:
Merge branch 'main' into main

Signed-off-by: Artemiy Degtyarev <chereshnevvv4real@gmail.com>

chrshnv committed:
Polishing.

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

New hasNext implementation.

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

Polishing

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

Remove unused nullable annotation and unused params.

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

Use already created methods.

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

mp911de commented:
FTR, we haven't forgotten you. We just have other priorities to consider at the moment.

chrshnv committed:
Fix tests.

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

chrshnv committed:
Add 'authored by'.

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

chrshnv committed:
Polishing

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>

chrshnv committed:
Using standard spring methods to obtain sort direction

Signed-off-by: Artemiy Chereshnevvv <chereshnevvv4real@gmail.com>
